engine/source/assets/ShaderAssetManager.cpp

```diff
@@ -0,0 +1,239 @@
+        }
+        if (stages == nc::ShaderStages::None)
+        {
+            throw nc::NcError("No valid shader stages found.");
+        }
+
+        return nc::DescriptorManifest
+        {
+            .setIndex = set,
+            .slotIndex = binding,
+            .descriptorType = descriptorType,
+            .shaderStages = stages
+        };
+    }
+
+    auto ReadShader(const std::string& path, const std::string uid) -> nc::ShaderView
+    {
+        const auto ncaString = (std::filesystem::path(path) / std::filesystem::path("shader.nca")).string();
+
+        if (!HasValidAssetExtension(ncaString)) throw nc::NcError("Invalid extension: " + ncaString);
+
+        std::ifstream file{ncaString};
+        if (!file.is_open()) throw nc::NcError("Could not open file: " + ncaString);
+
+        std::string vertexShaderPath;
+        file >> vertexShaderPath;
+        vertexShaderPath = (std::filesystem::path(path) / std::filesystem::path(vertexShaderPath)).string();
+
+        std::string fragmentShaderPath;
+        file >> fragmentShaderPath;
+        fragmentShaderPath = (std::filesystem::path(path) / std::filesystem::path(fragmentShaderPath)).string();
+
+        auto vertexByteCode = ReadSpirvIntoBytes(vertexShaderPath);
+        auto fragmentByteCode = ReadSpirvIntoBytes(fragmentShaderPath);
+
+        auto shaderView = nc::ShaderView
+        {
+            .uid ="",
+            .vertexByteCode = vertexByteCode,
+            .fragmentByteCode =  fragmentByteCode,
+            .descriptors = {}
+        };
+
+        uint32_t descriptorCount;
+        file >> descriptorCount;
+
+        for (uint32_t i = 0; i < descriptorCount; ++i)
+        {
+            shaderView.descriptors.emplace_back(ReadDescriptor(file));
+        }
+
+        shaderView.uid = uid;
+
+        return shaderView;
+    }
+
+}
+
+namespace nc
+{
+    ShaderAssetManager::ShaderAssetManager(const std::string& assetDirectory)
+        : m_shaderViews{},
+          m_assetDirectory{assetDirectory}
+    {
+    }
+
+    bool ShaderAssetManager::Load(const std::string& path, bool isExternal)
+    {
+        const auto fullPath = isExternal ? path : m_assetDirectory + path;
+
+        auto shaderView = ReadShader(fullPath, path);
+
+        if (IsLoaded(fullPath))
+        {
+            return false;
+        }
+
+        for (const auto& manifest : m_shaderViews)
+        {
+            if (manifest.uid == shaderView.uid)
+            {
+                return false;
+            }
+        }
+
+        if (shaderView.vertexByteCode.empty() || shaderView.fragmentByteCode.empty())
+        {
+            return false;
+        }
+
+        m_shaderViews.emplace_back(std::move(shaderView));
+
+        return true;
+    }
+
+    bool ShaderAssetManager::Load(std::span<const std::string> paths, bool isExternal)
+    {
+        auto atLeastOneFailure = false;
+        for (const auto& path : paths)
+        {
+            if (IsLoaded(path))
```

Comment:
No need to check `IsLoaded()` here since the single path version also has to check.